### PR TITLE
Support controlling interface and member visibility via preferences

### LINKED ISSUE #7626: Implement pref-controlled interfaces/methods/attributes in WebIDL
We want to be able to hid the MouseEvent::which attribute when `dom.mouseevent.which.enabled` pref is false, for example.

For prior art we can steal from Gecko, see PropertyDefiner.getControllingCondition and PropertyDefiner.generatePrefableArray, as well as any other code that uses `getExtendedAttribute("Pref")` in http://mxr.mozilla.org/mozilla-central/source/dom/bindings/Codegen.py .

Code: components/script/dom/bindings/codegen/CodegenRust.py

### PR DESCRIPTION
**Pull request by jdm**
This was easier to throw together than per-attribute/method support, and it gets rid of some nonstandard properties from our globals.

Fixes #7626.

r? @Ms2ger

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/10081)

### CODE REVIEW

**Comment by jdm** [comment]
I discovered that in the case of:
```
[Pref="foo"]
interface Parent {}

interface Child : Parent {}
```
`Parent` still gets defined in the current implementation even if `foo` is false. What should we do in this case?

**Comment by nox** [comment]
Do you mean that `Child` still gets defined? Anyway whether it is parent or child I don't think either of them should be visible.

**Comment by jdm** [comment]
No, I mean that window.Parent is still defined even if the preference controlling it's visibility is false.

**Comment by nox** [comment]
Well that's obviously wrong, no?

**Comment by larsbergstrom** [comment]
:+1: to `[Pref="dom.bluetooth.enabled"], and off by default.

cc @dati91

**Comment by jdm** [comment]
This is ready for review.

**Comment by jdm** [comment]
This is still ready for review; I just got carried away and implemented the Pref annotation for members of interfaces as well.

**Comment by paulrouget** [comment]
Can we get a review on that?

**Comment by jdm** [comment]
This has been ignored for more than a month. I would really appreciate if someone would take responsibility for reviewing my work :(

**Comment by nox** [comment]
On it.

Could you give more details about what you meant by "This was easier to throw together than per-attribute/method support"?

**Comment by nox** [comment]
@jdm I think you are cursed.

capture d ecran 2016-04-29 a 11 05 07

**Comment by pkaminski** [comment]
Hey @jdm, how did you make that happen? Reviewable thinks that somehow there's a discussion attached to a file that doesn't exist in the review, hence **boom**.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10081#-:-KGWI4XbnHx8lCjBoHw2:-676619027)*

**Comment by pkaminski** [comment]
I mean @nox, sorry.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10081#-:-KGWIcHs5-6zOnNeEjxP:-639219013)*

**Comment by nox** [comment]
@pkaminski I just clicked on the Reviewable button and it failed to load.

@jdm I don't get the purpose of avoiding `JS_DefineProperties` and friends, they just loop over items and call `JS_DefineProperty` sequentially. Couldn't we just rewrite `JS_DefineProperties` ourselves, putting a new field in `JSPropertySpec` or a similar structure?

File: components/script/dom/bindings/codegen/CodegenRust.py [hunk 1]
```diff
@@ -2414,10 +2550,10 @@ def definition_body(self):
         for arrayName in self.properties.arrayNames():
             array = getattr(self.properties, arrayName)
             if arrayName == "consts":
-                if array.length():
-                    properties[arrayName] = array.variableName()
+                if not array.length():
+                    properties[arrayName] = "EMPTY_CONSTANTS"
                 else:
-                    properties[arrayName] = "&[]"
+                    properties[arrayName] = array.variableName()
```
Comment (nox) [inline components/script/dom/bindings/codegen/CodegenRust.py]: Please revert this. It worked before and you didn't even document `EMPTY_CONSTANTS`.

**Comment by pkaminski** [comment]
Ah, this was the review I had reset a while back.  You had a draft against the old review, which blew Reviewable due to the dangling reference.  That's why you're not supposed to reset reviews if there's been any work done on them.  :)  Anyway, I deleted your old draft so you should be fine now.

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10081#-:-KGWJJE6Eq5V8YlHzKXo:166738921)*

**Comment by jdm** [comment]
Consts and statics can't contain slice expressions with ranges, which is why I had to reimplement the slice representation inside of `Prefable`. I don't see any way around this.

**Comment by jdm** [comment]
@nox I rewrote it to create slices of slices of the JS specs as desired. I agree that the result is nicer to reason about!

**Comment by nox** [comment]
Very nice! Just a small commenting remark and then you can r=me when you reply or address it.

-S-awaiting-review +S-awaiting-answer

---

Reviewed 2 of 2 files at r1, 15 of 15 files at r2, 1 of 1 files at r3, 1 of 1 files at r4, 3 of 3 files at r5, 8 of 8 files at r6, 3 of 3 files at r7, 27 of 27 files at r8, 13 of 15 files at r9, 1 of 1 files at r10, 1 of 1 files at r11, 8 of 8 files at r13, 3 of 4 files at r14, 1 of 1 files at r15.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[components/script/dom/bindings/interface.rs, line 66 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10081#-KGkUpyhEm1gKuKas-Wu:-KGkUpyhEm1gKuKas-Wv:1087066808) ([raw file](https://github.com/servo/servo/blob/dd80a75a78198a16c7aa9cc001742bfb0c4d84f5/components/script/dom/bindings/interface.rs#L66)):*
Please revert this and just use an empty slice directly wherever needed.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 1344 \[r15\]](https://reviewable.io:443/reviews/servo/servo/10081#-KGkVQme4cOq1QIh698Z:-KGkVQme4cOq1QIh698_:-212031085) ([raw file](https://github.com/servo/servo/blob/d2d383894ad311bd45f7c9aa8c658cacd05d5f85/components/script/dom/bindings/codegen/CodegenRust.py#L1344)):*
Doesn't Python define that itself?

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 1413 \[r15\]](https://reviewable.io:443/reviews/servo/servo/10081#-KGkVWnI1KEeFshKhI4K:-KGkVWnI1KEeFshKhI4L:458350648) ([raw file](https://github.com/servo/servo/blob/d2d383894ad311bd45f7c9aa8c658cacd05d5f85/components/script/dom/bindings/codegen/CodegenRust.py#L1413)):*
Ain't that comment false now that you rewrote everything to use multiple slices?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10081#-:-KGkW3Znq7Rtr04UUM5m:-2044119631)*

**Comment by jdm** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[components/script/dom/bindings/interface.rs, line 66 \[r5\]](https://reviewable.io:443/reviews/servo/servo/10081#-KGkUpyhEm1gKuKas-Wu:-KGlolugNNbTsgjsbHGP:-1814847148) ([raw file](https://github.com/servo/servo/blob/dd80a75a78198a16c7aa9cc001742bfb0c4d84f5/components/script/dom/bindings/interface.rs#L66)):*
This was a comment on an old revision.

---

*[components/script/dom/bindings/codegen/CodegenRust.py, line 1344 \[r15\]](https://reviewable.io:443/reviews/servo/servo/10081#-KGkVQme4cOq1QIh698Z:-KGlndo59xF8mVqMIwOi:1778814349) ([raw file](https://github.com/servo/servo/blob/d2d383894ad311bd45f7c9aa8c658cacd05d5f85/components/script/dom/bindings/codegen/CodegenRust.py#L1344)):*
From https://docs.python.org/2/reference/datamodel.html#object.__ne__, `Accordingly, when defining __eq__(), one should also define __ne__() so that the operators will behave as expected`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10081)*

**Comment by jdm** [comment]
@bors-servo: retry
* git

**Comment by ConnorGBrewster** [comment]
@bors-servo retry
 * Intermittent

**Pull request merged**